Note configurations are shared across all files in a module by instance given a uniquely-identifying opaque ID (#3357)
sass/language: spec/modules.md

### Configuration

A *configuration* is a map from [variable](variables.md) names to SassScript

> values. An *empty configuration* contains no entries. Each *configuration* has
> a uniquely-identifying opaque ID.

Reviewer: Nit: I'd include the ID in the first sentence, since it's (now) part of the structure of a configuration:

```suggestion
values and an opaque ID. An *empty configuration* contains no entries.
```

I'd omit "uniquely-identifying" because this is so far just describing the _structure_ of a configuration, not its _semantics_.

Each *configuration* has a uniquely-identifying opaque ID, sharing the same

> instance across all [source files][] in a [module](#module).

Reviewer: The first half of this is just reiterating the previous paragraph, and it's not clear exactly which new configurations the second part applies to. I'd say instead something like, "Each new configuration has a unique ID by default." Then in the text of the spec, you can say something like "Create a new configuration with the same ID as before" if you want them to share identities.

Author: done, I think.. :)

* If `file` is null, throw an error.

* If `file` has already been [executed] by the [Loading a Module] procedure:

  [executed]: spec.md#executing-a-file
  [Loading a Module]: #loading-a-module

  * If `config` is not empty and has a different ID than the configuration that

>     was passed the first time `file` was executed, throw an error.

Reviewer: "was executed by the Loading a Module procedure"

  [executed]: spec.md#executing-a-file

>   * If `config` has a different ID than before and is not empty, throw an error.

Reviewer: I think "than the configuration that was passed the first time `file` was executed" would be clearer.

Also, this is a pre-existing issue, but you might as well fix it while you're here: I think the previous bullet point should say "If `file` has already been executed **by the Loading a Module procedure**", since it's valid to load a file with `@import` and also with `@use`.

  [executed]: spec.md#executing-a-file

>   * If `config` has a different ID than before and is not empty, throw an error.

Reviewer: Currently, nothing in this spec says to create a config with anything other than a unique ID, so this is really just saying "if `config` is a different object than the previous one". If that's all we need to specify, then we can do away with the "opaque ID" concept entirely and just say that directly. But if we want to consider different configuration objects to be "the same"—for example if they're created with `@forward as` or `@forward with`—we need to explicitly say "create a new configuration with the same ID as ..." when they're created.

Author: I have added a note on `ForwardWithArgument` but it's a bit far so I added a note below, so info is not too far apart

  > This disallows circular `@use`s, which ensures that modules can't be used
  > until they're fully initialized.

* Otherwise

>   * Keep track of which `config` was used for the `file` by its opaque ID.

Reviewer: You don't need to explicitly specify when and where to track stuff in the spec—you can just refer to values and, as long as it's unambiguously specified what you're referring to, it's up to the implementation to figure out how to actually ensure the information is available when necessary.

In other words, specs only need to say _what to do_, they don't have to be concerned with _how to do it_.
